Add cluster reinstallation documentation (#238)
stolostron/siteconfig: docs/reinstallation/README.md

## Cluster Reinstallation Workflow  
Follow these steps to perform a cluster reinstallation:

1. **Enable the SiteConfig Operator reinstallation service**: If not already enabled, update the `siteconfig-operator-configuration` ConfigMap.
2. **Label resources for preservation**.
3. **Update the `ClusterInstance` resource**: Set `spec.reinstall.preservationMode`, then trigger reinstallation by assigning a new, unique value to `spec.reinstall.generation`. Apply any other necessary updates.
4. **Apply the changes** to initiate the reinstallation process.
5. **Monitor the reinstallation progress** via ClusterInstance `status.reinstall.conditions`:
    - The SiteConfig Operator validates the reinstallation request.
    - The operator preserves labeled data and removes existing installation manifests.
    - If resources were preserved, they are restored, completing the reinstallation process.
6. **The SiteConfig Operator regenerates new installation manifests** from the referenced templates.

> 7. **Monitor for provisioning completion**.

Author: Good question! These steps are part of the normal provisioning flow and not the reinstall handling flow.

## Enabling Cluster Reinstallation in the SiteConfig Operator

Cluster reinstallation must be explicitly enabled in the `siteconfig-operator-configuration` ConfigMap.

### Checking the Current Configuration  
By default, reinstallation is disabled. Verify the configuration:

```yaml
apiVersion: v1  
kind: ConfigMap  
metadata:  
  name: siteconfig-operator-configuration  
  namespace: <namespace>  
data:  
  allowsReinstalls: false  
  ...  
```

> - `<namespace>`: The namespace where the SiteConfig Operator is installed.  

Author: Not as such. If a standalone deployment - it will be `siteconfig-operator`. If an ACM deployment, most likely it will be `rhacm` or `open-cluster-management`.

### Verifying the Cluster Reinstallation

1. Ensure that the reinstallation request is being processed:  
```sh
oc get clusterinstance clusterinstance-example -n some-namespace -o json | jq -r '.status.reinstall.conditions[] | select(.type=="ReinstallRequestProcessed")'
{
  "type": "ReinstallRequestProcessed"
  "reason": "InProgress",
  "status": "False",
  ...
}
```

2. Ensure that the cluster reinstallation request was validated successfully:
```sh
oc get clusterinstance clusterinstance-example -n some-namespace -o json | jq -r '.status.reinstall.conditions[] | select(.type=="ReinstallRequestValidated")'
{
  "type": "ReinstallRequestValidated"
  "reason": "Completed",
  "status": "True",
  ...
}
```

3. Check that the cluster identity data are preserved if `spec.reinstall.preservationMode: All | ClusterIdentity`:
```sh
oc get clusterinstance clusterinstance-example -n some-namespace -o json | jq -r '.status.reinstall.conditions[] | select(.type=="ReinstallPreservationDataBackedup")'
{
  "type": "ReinstallPreservationDataBackedup"
  "reason": "Completed",
  "status": "True",
  ...
}
```

> 4. (optional) Ensure that the cluster identity data are detected if `spec.reinstall.preservationMode: All | ClusterIdentity`:

Reviewer: I'm also a bit torn on using either "if" or "when"

5. Ensure that the installation manifests are deleted. This can take several minutes to complete.
```sh
oc get clusterinstance clusterinstance-example -n some-namespace -o json | jq -r '.status.reinstall.conditions[] | select(.type=="ReinstallRenderedManifestsDeleted")'
{
  "type": "ReinstallRenderedManifestsDeleted"
  "reason": "Completed",
  "status": "True",
  ...
}
```

6. If `preseservationMode` is `All` or `ClusterIdentity`, ensure that the data preserved earlier are restored.
```sh
oc get clusterinstance clusterinstance-example -n some-namespace -o json | jq -r '.status.reinstall.conditions[] | select(.type=="ReinstallPreservationDataRestored")'
{
  "type": "ReinstallPreservationDataRestored"
  "reason": "Completed",
  "status": "True",
  ...
}
```

7. Once all the above mentioned steps are verified, ensure that reinstallation request completes successfully.
```sh
oc get clusterinstance clusterinstance-example -n some-namespace -o json | jq -r '.status.reinstall.conditions[] | select(.type=="ReinstallRequestProcessed")'
{
  "type": "ReinstallRequestProcessed"
  "reason": "Completed",
  "status": "True",
  ...
}
```

## Advanced Topics

### Image-Based Break/Fix (IBBF)

> The Image-Based Break/Fix (IBBF) feature simplifies Single Node OpenShift (SNO) hardware replacement, minimizing downtime while preserving the cluster’s original identity. It retains critical cluster details, including identifiers, cryptographic keys (such as `kubeconfig`), and authentication credentials, enabling the replacement node to seamlessly assume the identity of the failed hardware.

Reviewer: ```suggestion
The Image-Based Break/Fix (IBBF) feature simplifies single-node OpenShift (SNO) hardware replacement by minimizing downtime while preserving the cluster’s original identity. The IBBF feature retains critical cluster details, including identifiers, cryptographic keys, such as `kubeconfig`, and authentication credentials, enabling the replacement node to seamlessly assume the identity of the failed hardware.
```

   - [Additional Reinstallation Tracking Information](#additional-reinstallation-tracking-information)  
7. [Advanced Topics](#advanced-topics)  
   - [Image-Based Break/Fix (IBBF)](#image-based-breakfix-ibb)

</details>

## Overview
The SiteConfig operator simplifies OpenShift cluster reinstallation through the `ClusterInstance` API while preserving critical configuration data.

With a GitOps-compatible, declarative approach, users can trigger reinstallations by updating the `ClusterInstance` resource. The operator also includes a backup and restore mechanism for `Secret` and `ConfigMap` resources, ensuring essential cluster data, such as authentication credentials and configuration resources, remains intact.

### Cluster Identity Preservation
Cluster reinstallation supports both Ssingle-node OpenShift (SNO) and multi-node OpenShift (MNO) clusters. However, cluster identity preservation is only supported for Single Node OpenShift clusters installed using the Image Based Install (IBI) provisioning method.

## Cluster Reinstallation Workflow  
To perform a cluster reinstallation, complete the following steps:

> 1. **Enable the SiteConfig operator reinstallation service**: If not already enabled, update the `siteconfig-operator-configuration` `ConfigMap` resource.

Author: > Or maybe make this overview a bit more minimal if it's just intended to be a high level description of the steps and leave the particular details for the later section?
+1

> 1. **Enable the SiteConfig operator reinstallation service**: If not already enabled, update the `siteconfig-operator-configuration` `ConfigMap` resource.

Author: @carbonin we decided to incorporate the IBBF doc under the same `README.md`. This avoids duplication of content.

2. **Label resources for preservation**.
3. **Update the `ClusterInstance` resource**: 

>    - Set the `spec.reinstall.preservationMode` field.

Author: Only steps 1 and 2 can be performed ahead of time. Will refactor the docs to capture this.

## Labeling Resources for Preservation
To retain essential cluster configuration data after reinstallation, the SiteConfig operator provides a backup and restore mechanism for `ConfigMap` and `Secret` resources within the `ClusterInstance` namespace.

To preserve specific resources, apply the appropriate preservation label as described in the [preservation document](#preservation.md). The SiteConfig operator backs up resources based on the following labels:

- `siteconfig.open-cluster-management.io/preserve: "<arbitrary-value>"`: Backs up both non-identity-related and cluster identity-related resources.
- `siteconfig.open-cluster-management.io/preserve: "cluster-identity"`: Backs up cluster identity-related resources.

To label a resource for preservation, run the following command:

```sh
oc label configmap <your_configmap> "siteconfig.open-cluster-management.io/preserve="
```

The `ClusterInstance` CR is updated with the correct `preservationMode` corresponding to the applied labels. The following preservation modes are supported:

- `None`: No data is backed up.
- `All`: Backs up all Secrets and ConfigMaps that have the label `siteconfig.open-cluster-management.io/preserve`.

> - `ClusterIdentity`: Backs up only Secrets and ConfigMaps labeled with `siteconfig.open-cluster-management.io/preserve: "cluster-identity"`.  

Reviewer: I have the following descriptions from an older draft: 
`siteconfig.open-cluster-management.io/preserve: " "`:  Use this label to back up any resources that are not related to cluster identity.
`siteconfig.open-cluster-management.io/preserve: “cluster-identity”`: Use this label to back up resources that are related to cluster identity.

Feel free to use it if it's still applicable.

> - `ClusterIdentity`: Backs up only Secrets and ConfigMaps labeled with `siteconfig.open-cluster-management.io/preserve: "cluster-identity"`.  

Author: Will elaborate on this. Great feedback!

## Initiating a Cluster Reinstallation

To start the reinstallation, update the `ClusterInstance` resource with a new `spec.reinstall.generation` value.

### Updating the `ClusterInstance` Resource  

Modify the `spec.reinstall.generation` value in the `ClusterInstance` resource:

```yaml
apiVersion: siteconfig.open-cluster-management.io/v1alpha1  
kind: `ClusterInstance`  
metadata:  
  name: clusterinstance-example  
  namespace: some-namespace  
spec:  
  reinstall:  
    generation: "unique-generation-string"  
    preservationMode: "None | All | ClusterIdentity"  
```

When defining the `spec.reinstall` object, you can update the following fields:

- `spec.extraAnnotations`
- `spec.extraLabels`
- `spec.suppressedManifests`
- `spec.pruneManifests`
- `spec.nodes/<node-index>/extraAnnotations`
- `spec.nodes/<node-index>/extraLabels`
- `spec.nodes/<node-index>/suppressedManifests`
- `spec.nodes/<node-index>/pruneManifests`

> :information_source: `<node-index>` represents the index of the `NodeSpec` object.

Reviewer 2: Is this node index meaningful to users or would it be easier to just omit it here?

Author: Nodes are a list of NodeSpec objects - the `node-index` here aims to draw attention to the user on which node is being "updated".

### Applying the Changes  
Use one of the following methods:

> - **GitOps Approach**: Commit and push the updated `ClusterInstance` resource to your Git repository.

Reviewer 2: Do we need to call this out anywhere? Nothing else in this repo talks about or depends on gitops, right?

Author: > Do we need to call this out anywhere? Nothing else in this repo talks about or depends on gitops, right?

I wrote these docs with the intention to easily translate them into KB articles and downstream docs. I plan to work with @amolnar-rh to add a similar doc on deploying a cluster with the ClusterInstance. Also note that we do have a section dedicated to configuring ArgoCD. So maybe it's fine 🤷‍♂️

4. [Optional. Monitor the reinstallation progress.](#4-optional-monitoring-the-cluster-reinstallation)
5. [Verify that the cluster is provisioned and available.](#5-verify-cluster-availability)

You can complete step 1. and 2. well before initiating the cluster reinstallation.

### 1. Enabling cluster reinstallation in the SiteConfig operator

You must explicitly enable cluster reinstallation in the `siteconfig-operator-configuration` `ConfigMap` resource, which you can complete well before initiating the process.

#### Checking the current configuration  
By default, reinstallation is disabled. You can check the current configuration by completing the following steps:

Create the `NAMESPACE` environment variable for the `siteconfig-operator-configuration` `ConfigMap` resource. Run the following command:

```sh
NAMESPACE=<namespace>
```

Verify the current configuration by running the following command:

```sh
oc get configmap siteconfig-operator-configuration -n $NAMESPACE -o yaml
```

```yaml
apiVersion: v1  
kind: `ConfigMap`  
metadata:  
  name: siteconfig-operator-configuration  
  namespace: <namespace>  
data:  
  allowsReinstalls: false  
  ...  
```

> - `<namespace>`: The namespace where the SiteConfig operator is installed.  

Reviewer 2: Can't this be found using the downwardAPI on the deployment? Why does it need to be provided?

Reviewer: I'm wondering if I missed this and it should say "The namespace must match the namespace where the SiteConfig operator is installed."? @sakhoury

Author: @carbonin Good point! The Downward API can expose the namespace to the operator itself at runtime, but this instruction is written for the user, not for the operator. The `oc get configmap` ... command is meant to be run manually by the user from a shell which doesn't have automatic access to the operator's namespace unless explicitly provided.

Since the SiteConfig operator can be installed in different namespaces (`rhacm`, `open-cluster-management`, or `siteconfig-operator`), we prompt the user to supply it to ensure the command works correctly in their specific environment.

For more information, see [Preservation Modes](preservation.md#preservation-modes).

### 3. Initiating a cluster reinstallation

To start the reinstallation, update the `ClusterInstance` resource with a new `spec.reinstall.generation` value.

#### Updating the ClusterInstance resource  

Modify the `spec.reinstall.generation` value in the `ClusterInstance` resource:

```yaml
apiVersion: siteconfig.open-cluster-management.io/v1alpha1  
kind: `ClusterInstance`  
metadata:  
  name: clusterinstance-example  
  namespace: some-namespace  
spec:  
  reinstall:  
    generation: "unique-generation-string"  
    preservationMode: "<your-preservation-mode>"  
```

:information_source: Ensure you set the appropriate preservation mode when you modify the `ClusterInstance` resource. `"None"`, `"All"`, or `"ClusterIdentity"` are valid values.

When defining the `spec.reinstall` object, you can modify the following additional fields in the `ClusterInstance` resource:

- `spec.extraAnnotations`
- `spec.extraLabels`
- `spec.suppressedManifests`
- `spec.pruneManifests`
- `spec.nodes/<node-id>/extraAnnotations`
- `spec.nodes/<node-id>/extraLabels`
- `spec.nodes/<node-id>/suppressedManifests`

> - `spec.nodes/<node-id>/pruneManifests`

Author: Thanks @trewest! @amolnar-rh please add the following items:
```
/nodes/<node-id>/bmcAddress,
/nodes/<node-id>/bootMACAddress,
/nodes/<node-id>/nodeNetwork/interfaces/macAddress,
/nodes/<node-id>/rootDeviceHints,
```

> - `spec.nodes/<node-id>/pruneManifests`

Reviewer: Done

1. [Enable the SiteConfig operator reinstallation service.](#enabling-cluster-reinstallation-in-the-siteconfig-operator)
2. [Initiate a cluster reinstallation.](#initiating-the-image-based-breakfix-cluster-reinstallation)
   - Set `spec.reinstall.preservationMode: "ClusterIdentity"`.
   - Update the `spec.nodes` object with the changed hardware information.
   - :information_source: Note that the Image Based Install Operator automatically labels cluster identity resources.
3. [Optional. Monitor the reinstallation progress.](#4-optional-monitoring-the-cluster-reinstallation)
   - During cluster provisioning, the operators consuming the installation manifests provision the cluster using the newly generated manifests.
     :information_source: Note that the [Image Based Install Operator](https://github.com/openshift/image-based-install-operator) detects the preserved cluster identity data and incorporates it into the configuration ISO image.
4. [Verify that the cluster is provisioned and available.](#5-verify-cluster-availability)
   - Use the `kubeconfig` that is associated with the failed hardware to access the reinstalled spoke cluster.

> ## Initiating the Image-Based Break/Fix cluster reinstallation

Author: Should this be a level 2 heading?

Reviewer: Nope, it should be level 4! Thanks for catching it, fixed it just now

8. [Advanced topics](#advanced-topics)  
   - [Image-Based Break/Fix](#image-based-breakfix-for-single-node-openshift-hardware-replacement)

</details>

## Overview
The SiteConfig operator simplifies OpenShift cluster reinstallation through the `ClusterInstance` API while preserving critical configuration data.

With a GitOps-compatible, declarative approach, users can trigger reinstallations by updating the `ClusterInstance` resource. The operator also includes a backup and restore mechanism for `Secret` and `ConfigMap` resources, ensuring essential cluster data, such as authentication credentials and configuration resources, remains intact.

### Cluster identity preservation
Cluster reinstallation supports both single-node OpenShift and multi-node OpenShift clusters. However, cluster identity preservation is only supported for single-node OpenShift clusters installed using the Image Based Install provisioning method.

## Cluster reinstallation workflow  
To perform a cluster reinstallation, complete the following steps:

1. [Enable the SiteConfig operator reinstallation service.](#1-enabling-cluster-reinstallation-in-the-siteconfig-operator)

> 2. [Label resources for preservation.](#2-labeling-resources-for-preservation)

Reviewer 3: I believe this step is also optional?

3. [Initiate a cluster reinstallation.](#3-initiating-a-cluster-reinstallation)
4. [Optional. Monitor the reinstallation progress.](#4-optional-monitoring-the-cluster-reinstallation)
5. [Verify that the cluster is provisioned and available.](#5-verify-cluster-availability)

> You can complete step 1. and 2. well before of initiating the cluster reinstallation.

Reviewer 3: ```suggestion
You can complete step 1. and 2. well before initiating the cluster reinstallation.
```

### 1. Enabling cluster reinstallation in the SiteConfig operator

You must explicitly enable cluster reinstallation in the `siteconfig-operator-configuration` `ConfigMap` resource, which you can complete well before initiating the process.

#### Checking the current configuration  
By default, reinstallation is disabled. You can check the current configuration by completing the following steps:

Create the `NAMESPACE` environment variable for the `siteconfig-operator-configuration` `ConfigMap` resource. Run the following command:

```sh
NAMESPACE=<namespace>

> ```

Reviewer 3: Why do I need to define an env var for this process?

Author: Defining the `NAMESPACE` variable ensures clarity, reduces manual errors and decouples the namespace value (which can vary between `rhacm`, `open-cluster-management`, or `siteconfig-operator` depending on how the operator was installed) making the instructions flexible.

#### Enabling cluster reinstallation  

Create the `NAMESPACE` environment variable for the `siteconfig-operator-configuration` `ConfigMap` resource. Run the following command:

```sh
NAMESPACE=<namespace>
```

Update the `ConfigMap` resource by running the following command:

```sh
oc patch configmap siteconfig-operator-configuration \  
  -n $NAMESPACE \  
  --type=json \  
  -p '[{"op": "replace", "path": "/data/allowsReinstalls", "value": "true"}]'  
```

Verify the update by running the following command:

```sh
oc get configmap siteconfig-operator-configuration -n $NAMESPACE -o yaml  
```

### 2. Labeling resources for preservation

To retain essential cluster configuration data after reinstallation, the SiteConfig operator provides a backup and restore mechanism for `ConfigMap` and `Secret` resources within the `ClusterInstance` namespace.

To preserve specific resources, apply the appropriate preservation label as described in the [preservation document](#preservation.md). You can label your resources well before initiating the reinstallation.

The SiteConfig operator backs up resources based on the following labels:

> - `siteconfig.open-cluster-management.io/preserve: "<arbitrary-value>"`: Backs up both non-identity-related and cluster identity-related resources.

Reviewer 3: Wording is misleading. The label does not back anything up. CRs with an arbitrary value will be backed up only under the All preservation mode

> - `siteconfig.open-cluster-management.io/preserve: "cluster-identity"`: Backs up cluster identity-related resources.

Reviewer 3: CRs with this label will be backed up under either the All or ClusterIdentity preservation Mode

- `ClusterIdentity`: Backs up only `Secret` and `ConfigMap` resources labeled with `siteconfig.open-cluster-management.io/preserve: "cluster-identity"`. If no resources are labeled, the operator stops the reinstallation and displays an error message.

For more information, see [Preservation Modes](preservation.md#preservation-modes).

### 3. Initiating a cluster reinstallation

To start the reinstallation, update the `ClusterInstance` resource with a new `spec.reinstall.generation` value.

#### Updating the ClusterInstance resource  

Modify the `spec.reinstall.generation` value in the `ClusterInstance` resource:

```yaml
apiVersion: siteconfig.open-cluster-management.io/v1alpha1  
kind: `ClusterInstance`  
metadata:  
  name: clusterinstance-example  
  namespace: some-namespace  
spec:  
  reinstall:  
    generation: "unique-generation-string"  
    preservationMode: "<your-preservation-mode>"  
```

:information_source: Ensure you set the appropriate preservation mode when you modify the `ClusterInstance` resource. `"None"`, `"All"`, or `"ClusterIdentity"` are valid values.

> When defining the `spec.reinstall` object, you can update the following fields:

Reviewer 3: ```suggestion
When defining the `spec.reinstall` object, the following additional fields in the ClusterInstance may be modified:
```

- `ReinstallRenderedManifestsDeleted` – Monitors the deletion of rendered manifests associated with the `ClusterInstance` resource.
- `ReinstallPreservationDataRestored` – Tracks the restoration status of preserved data.

The`status.reinstall` field provides further information about the reinstallation process with the following fields:

- `InProgressGeneration` – Identifies the active generation being processed for reinstallation.
- `ObservedGeneration` – Indicates the last successfully processed reinstallation request.
- `RequestStartTime` – Indicates the time when the reinstallation request was initiated.
- `RequestEndTime` – Indicates the time when the reinstallation process was completed.
- `History` – Displays past reinstallation attempts, including generation details, timestamps, and specification changes to the `ClusterInstance` resource.

For more information about the reinstallation conditions, see [Cluster reinstallation status conditions](status_conditions.md).

To monitor the cluster reinstallation progress, complete the following steps:

1. Verify that the reinstallation request is being processed:  

    ```sh

>     oc get clusterinstance clusterinstance-example -n some-namespace -o json | jq -r '.status.reinstall.conditions[] | select(.type=="ReinstallRequestProcessed")'

Reviewer 3: Can use the oc tools 'custom-columns' to avoid dependence on the external jq tool (here and others):

```suggestion
    oc get clusterinstance cnfdf02 -o custom-columns=STATUS:'.status.conditions[?(@.type=="ReinstallRequestProcessed")].status'
```
